### PR TITLE
feat: added component query api

### PR DESCRIPTION
**Pull request by mcollovati**
The component query API that allows to scan the UI or a subtree for components matching given predicates,
and provides component wrappers for found elements.

Part of #1338

### CODE REVIEW

File: vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+     */
+    private static class LocatorSpec<T extends Component> {
+
+        public String id;
+        public String caption;
+        public String placeholder;
+        public String text;
+        public IntRange count = new IntRange(0, Integer.MAX_VALUE);
+        public Object value;
+        public String classes;
```
Comment (knoobie) [inline vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java:185]: Should there be a way to search for "contains class" or "has theme"?

Comment (mcollovati) [reply vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java:204]: Thanks for your feedback!
The implementation is still in progress. Additional query helpers will be added, but perhaps in other PRs.

**Review by caalador — CHANGES_REQUESTED** [review]

File: vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+import com.vaadin.flow.component.Component;
+import com.vaadin.testbench.unit.internal.LocatorKt;
+import com.vaadin.testbench.unit.internal.SearchSpec;
+
+/**
+ * Query class used for finding a component inside a given search context.
+ * 
+ * The search context is either the current {@link com.vaadin.flow.component.UI}
+ * instance which searches through the whole components tree, or a
```
[the remark this answers is not in the document]
Comment (mcollovati) [reply vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java]: fixed

File: vaadin-testbench-core/src/test/java/com/vaadin/testbench/unit/ComponentQueryTest.java [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+                "Expecting query to find TextField component, but got different instance");
+
+        ComponentQuery<Button> buttonQuery = select(Button.class);
+        Assertions.assertSame(button, buttonQuery.last().getComponent(),
+                "Expecting query to find Button component, but got different instance");
+
+    }
+
+    @Test
+    void last_multipleMatching_getsFirstComponent() {
```
[the remark this answers is not in the document]
Comment (mcollovati) [reply vaadin-testbench-core/src/test/java/com/vaadin/testbench/unit/ComponentQueryTest.java]: fixed

File: vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+            Class<E> componentType) {
+        return thenOn(1, componentType);
+    }
+
+    /**
+     * Gets a new {@link ComponentQuery} to search for given component type on
+     * the context of the matching component at given index for current query.
+     *
+     * Index is 1-based. Given a zero or negative index or an index higher than
+     * the actual number of components found results in an At
```
Comment (caalador) [inline vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java:139]:
```suggestion
     * the actual number of components found results in an
```

File: vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+                .orElseThrow(() -> new NoSuchElementException(
+                        "Cannot find component for current query"));
+    }
+
+    /**
+     * Executes the search against current context and returns the test wrapper
+     * for the component at given index.
+     *
+     * Index is 1-based. Given a zero or negative index or an index higher than
+     * the actual number of components found results in an At
```
Comment (caalador) [inline vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java:199]:
```suggestion
     * the actual number of components found results in an
```

**Review by caalador — COMMENTED** [review]
Apart for the 2 typos in javadoc is ok.

File: vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java (hunk 4 above)
Comment (mcollovati) [reply vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java]: fixed

**Review by caalador — APPROVED** [review]

**Pull request merged**